### PR TITLE
[REVIEW] Remove raw allocation from column_device_view

### PR DESCRIPTION
**Pull request by jrhemstad**
part of https://github.com/rapidsai/cudf/issues/4212

### CODE REVIEW

File: cpp/src/column/column_device_view.cu [hunk 1]
```diff
@@ -69,48 +69,65 @@ column_device_view::column_device_view( column_view source, ptrdiff_t h_ptr, ptr
...
+
+  // First calculate the size of memory needed to hold the
+  // child columns. This is done by calling extent()
+  // for each of the children.
+  auto get_extent = thrust::make_transform_iterator(
+      thrust::make_counting_iterator(0),
+      [&source](auto i) { return extent(source.child(i)); });
+
+  std::size_t const child_storage_bytes =
+      std::accumulate(get_extent, get_extent + num_children, 0);
```
Comment (codereport) [inline cpp/src/column/column_device_view.cu]:
```suggestion
      std::accumulate(get_extent, get_extent + num_children, std::size_t{0});
```

File: cpp/src/column/column_device_view.cu [hunk 2]
```diff
@@ -69,48 +69,65 @@ column_device_view::column_device_view( column_view source, ptrdiff_t h_ptr, ptr
...
+  // from CPU to device.
+  rmm::device_buffer* const child_storage = new rmm::device_buffer(child_storage_bytes, stream);
+  char* d_start = static_cast<char*>(child_storage->data());
+
+  auto deleter = [child_storage](column_device_view * v){
+    v->destroy();
+    delete child_storage;
+  };
+
+  using Deleter = std::function<void(column_device_view*)>;
```
Comment (codereport) [inline cpp/src/column/column_device_view.cu]: On line 72 you have the same type in the return type. Worth moving up to use in both places?

File: cpp/CMakeLists.txt [hunk 3]
```diff
@@ -112,6 +112,12 @@ set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} --expt-extended-lambda --expt-relaxed-
 # TODO: remove `no-maybe-unitialized` used to suppress warnings in rmm::exec_policy
 set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -Werror cross-execution-space-call -Xcompiler -Wall,-Werror,-Wno-error=deprecated-declarations")
 
+option(DISABLE_DEPRACATION_WARNING "Disable warnings generated from deprecated declarations." OFF)
```
Comment (harrism) [inline cpp/CMakeLists.txt]: Spelling.

```suggestion
option(DISABLE_DEPRECATION_WARNING "Disable warnings generated from deprecated declarations." OFF)
```

**Comment by davidwendt** [comment]
This looks good. Looks like this PR is dependent on #4542 ?
It probably does not have to be.

**Comment by jrhemstad** [comment]
> This looks good. Looks like this PR is dependent on #4542 ?
> It probably does not have to be.

Other way around. #4542 depends on this PR.

File: cpp/src/column/column_device_view.cu [hunk 4]
```diff
@@ -146,11 +161,13 @@ std::unique_ptr<mutable_column_device_view, std::function<void(mutable_column_de
...
 }
 
-size_type mutable_column_device_view::extent(mutable_column_view source) {
-  size_type data_size = sizeof(mutable_column_device_view);
-  for( size_type idx=0; idx < source.num_children(); ++idx )
-    data_size += extent(source.child(idx));
-  return data_size;
+std::size_t mutable_column_device_view::extent(mutable_column_view source) {
+  auto get_extent = thrust::make_transform_iterator(
+      thrust::make_constant_iterator(0),
```
Comment (harrism) [inline cpp/src/column/column_device_view.cu]: Should this be a counting iterator? Seems weird to always use child 0.

Comment (jrhemstad) [reply cpp/src/column/column_device_view.cu]: Ho boy, good catch. Auto-complete did me wrong.

**Comment by jrhemstad** [comment]
rerun tests

**Pull request merged**